### PR TITLE
bank-account: add benchmarks

### PR DESCRIPTION
**Pull request by tleen**
Given that locking seems to be the solution for the problem, I think
benchmarking could be a real interesting add-on to this specific
exercise.

### CODE REVIEW

**Comment by tleen** [comment]
This is as much for discussion of the benchmarks as anything. They may be too naive to be actually helpful in benchmarking.

**Comment by robphoenix** [comment]
I've not seen/used this before, but it certainly looks like you've implemented it correctly. 

**TLDR:** I'd question the benefit of throwing these benchmarks out there without relevant comments/documentation.

---

My only concern is that it may cause more confusion. Running the benchmarks gives me the following result:

```
> go test -bench=.       
BenchmarkAccountOperations-4             	10000000	       171 ns/op
BenchmarkAccountOperationsConcurrent-4   	 5000000	       284 ns/op
PASS
ok  	github.com/robphoenix/exercism-go/exercises/bank-account	3.640s
```

Thinking back to when I first spent time with benchmarks and concurrency in an exercism exercise, (I can't think which exercise now, I'll link to it if I find a suitable reference), here I would be confused as to why the concurrent operations are slower, `284 ns/op` vs. `171 ns/op`, because I would assume they would be faster. I realise this is down to my lack of understanding about concurrency.  I don't think adding benchmarks necessarily adds anything to the understanding of concurrency here, unless we can add some documentation that can explain these results or link to some helpful text, and these results may lead someone to think they're `doing it wrong`. Or maybe I'm just mistaken in my understanding of concurrency benchmarks, if so it's probably still useful to include an explanation to avoid others being similarly mistaken.

**Comment by tleen** [comment]
I would think they are slower because of the overhead of doing the concurrency compared to the not-complex nature of the actual operation that is being performed concurrently. If it was some crazy intensive task I think concurrency would be a performance improvement but here the operation is simple. The locks we are doing here act as a protective measure against the concurrency. The idea being for performance-based concurrency you want to make concurrent an operation that is so intensive that it justifies whatever overhead slowdown occurs with managing the concurrent resources.

Not really 100% sure either, don't have a lot of experience with concurrency in Go. :smile:

**Comment by robphoenix** [comment]
> I would think they are slower because of the overhead of doing the concurrency compared to the not-complex nature of the actual operation that is being performed concurrently.

I think if we include a comment along these lines, then this would be good to merge.

> don't have a lot of experience with concurrency in Go. 

That makes two of us :smile:

**Comment by tleen** [comment]
I'm not sure if these should be included, perhaps someone with more exp in this area can take a look? @petertseng?

The goal with this is to let people play with how they lock the vars to see if there is a performance impact. I would have to add more comments. Prob in the README.

**Comment by robphoenix** [comment]
I don't see a problem with including it as long as it's accompanied by some comments, much as you've explained in this thread @tleen

File: exercises/bank-account/bank_account_test.go [hunk 1]
```diff
@@ -296,6 +296,31 @@ func TestConcDeposit(t *testing.T) {
 	}
 }
 
+// The benchmark operations are here to allow you to try different
```
Comment (robphoenix) [inline exercises/bank-account/bank_account_test.go]: maybe change allow to help
allow doesn't feel quite right, this is more assistance than permission (nitpicky :smile: )

**Comment by robphoenix** [comment]
Just wondering if this needs to be 2 commits or if squashing them into a single commit would be better. I'm not so concerned but it might make for a cleaner commit history? Totally up to you. :+1:

**Comment by robphoenix** [comment]
**encourage** - even better!

File: exercises/bank-account/bank_account_test.go (hunk 1 above)
Comment (tleen) [inline exercises/bank-account/bank_account_test.go]: Taking another look I think "encourage" is the right word. 

(I appreciate the feedback, more feedback == better commits!)

**Comment by tleen** [comment]
:smile: I think one squashed commit: a benchmarks all or nothing.
